### PR TITLE
WIP: batch commit callbacks from all components in the render queue

### PR DESCRIPTION
**Pull request by developit**
Still one failing test, and we'll want to decide how best to alter the semantics of `options._commit`. Right now I have it attempting to emulate the old calling order, so _commit(vnode) gets called for every component with a dummy queue.

### CODE REVIEW

File: src/component.js [hunk 1]
```diff
@@ -220,21 +221,35 @@ const depthSort = (a, b) => a._vnode._depth - b._vnode._depth;
...
-			renderComponent(c);
-			if (rerenderQueue.length > renderQueueLength) {
+			root = renderComponent(c, commitQueue, refQueue) || root;
+			// If this WAS the last component in the queue, run commit callbacks *before* we exit the tight loop.
+			// This is required in order for `componentDidMount(){this.setState()}` to be batched into one flush.
+			// Otherwise, also run commit callbacks if the render queue was mutated.
+			if (renderQueueLength === 0 || rerenderQueue.length > renderQueueLength) {
+				commitRoot(commitQueue, root, refQueue);
+				commitQueue.length = 0;
+				refQueue.length = 0;
```
Comment (JoviDeCroock) [inline src/component.js:240]:
```suggestion
				refQueue.length = commitQueue.length = 0;
```

**Review by JoviDeCroock — APPROVED** [review]

**Review by marvinhagemeister — APPROVED** [review]
Let's remove the commented out code, then it's good to land.

File: src/component.js [hunk 2]
```diff
@@ -220,21 +221,34 @@ const depthSort = (a, b) => a._vnode._depth - b._vnode._depth;
...
 				// keep the order from top to bottom with those new items so we can handle them in a
 				// single pass
 				rerenderQueue.sort(depthSort);
+			} else if (root) {
+				if (options._commit) options._commit(root, EMPTY_ARR);
 			}
 		}
 	}
+	if (root) commitRoot(commitQueue, root, refQueue);
+	// if (root) commitRoot(commitQueue, root, refQueue);
```
Comment (developit) [inline src/component.js:251]:
```suggestion
```

File: src/render.js [hunk 3]
```diff
@@ -60,6 +60,9 @@ export function render(vnode, parentDom, replaceNode) {
 	);
 
 	// Flush all queued effects
+	// vnode._nextDom = undefined;
+	// if (options._commit) options._commit(vnode, EMPTY_ARR);
+	vnode._nextDom = undefined;
```
Comment (developit) [inline src/render.js:65]:
```suggestion
	vnode._nextDom = undefined;
```

File: compat/test/browser/useSyncExternalStore.test.js [hunk 4]
```diff
@@ -658,10 +658,11 @@ describe('useSyncExternalStore', () => {
...
 				store.set(1);
 			});
-			// Preact logs differ from React here cuz of how we do rerendering. We
-			// rerender subtrees and then commit effects so Child2 never sees the
-			// update to 1 cuz Child1 rerenders and runs its layout effects first.
-			assertLog([1, /*1,*/ 'Reset back to 0', 0, 0]);
+			// // Preact logs differ from React here cuz of how we do rerendering. We
+			// // rerender subtrees and then commit effects so Child2 never sees the
+			// // update to 1 cuz Child1 rerenders and runs its layout effects first.
+			// assertLog([1, /*1,*/ 'Reset back to 0', 0, 0]);
```
Comment (developit) [inline compat/test/browser/useSyncExternalStore.test.js:664]:
```suggestion
```

**Pull request merged**